packages/@aws-cdk/integ-runner/bin/integ-runner.ts

```diff
@@ -0,0 +1,196 @@
+        break;
+      case DiagnosticReason.TEST_FAILED:
+        logger.error('  %s - Failed!\n%s', diagnostic.testName, diagnostic.message);
+    }
+  });
+}
+
+async function main() {
+  const argv = yargs
+    .usage('Usage: integ-runner [TEST...]')
+    .option('list', { type: 'boolean', default: false, desc: 'List tests instead of running them' })
+    .option('clean', { type: 'boolean', default: true, desc: 'Skips stack clean up after test is completed (use --no-clean to negate)' })
+    .option('verbose', { type: 'boolean', default: false, alias: 'v', desc: 'Verbose logs' })
+    .option('dry-run', { type: 'boolean', default: false, desc: 'do not actually deploy the stack. just update the snapshot (not recommended!)' })
+    .option('update-on-failed', { type: 'boolean', default: false, desc: 'rerun integration tests and update snapshots for failed tests.' })
+    .option('force', { type: 'boolean', default: false, desc: 'Rerun all integration tests even if tests are passing' })
+    .option('parallel', { type: 'boolean', default: false, desc: 'run integration tests in parallel' })
+    .option('parallel-regions', { type: 'array', desc: 'if --parallel is used then these regions are used to run tests in parallel', nargs: 1, default: [] })
+    .options('directory', { type: 'string', default: 'test', desc: 'starting directory to discover integration tests' })
+    .argv;
+
+  // Use about half the advertised cores because hyperthreading doesn't seem to
+  // help that much, or we become I/O-bound at some point. On my machine, using
+  // more than half the cores actually makes it slower.
+  // Cap to a reasonable top-level limit to prevent thrash on machines with many, many cores.
+  const maxWorkers = parseInt(process.env.CDK_INTEG_MAX_WORKER_COUNT ?? '16');
+  const N = Math.min(maxWorkers, Math.max(1, Math.ceil(os.cpus().length / 2)));
+  const pool = workerpool.pool(path.join(__dirname, '../lib/workers/extract_worker.js'), {
+    maxWorkers: N,
+  });
+
+  // list of integration tests that will be executed
+  const testsToRun: IntegTestConfig[] = [];
+  const testsFromArgs: IntegTestConfig[] = [];
+  const parallelRegions = arrayFromYargs(argv['parallel-regions']);
+  const testRegions: string[] = parallelRegions ?? ['us-east-1', 'us-east-2', 'us-west-2'];
+  const runUpdateOnFailed = argv['update-on-failed'] ?? false;
+
+
+  try {
+    if (argv.list) {
+      const tests = await new IntegrationTests(argv.directory).fromCliArgs();
+      process.stdout.write(tests.map(t => t.fileName).join('\n') + '\n');
+      return;
+    }
+
+    if (argv._.length === 0) {
+      testsFromArgs.push(...(await new IntegrationTests(argv.directory).fromCliArgs()));
+    } else {
+      testsFromArgs.push(...(await new IntegrationTests(argv.directory).fromCliArgs(argv._.map(x => x.toString()))));
+    }
+
+    const diagnostics = new Array<Diagnostic>();
+    if (!argv.force) {
+      const requests = batchTests(testsFromArgs);
```

Comment:
(minor) there are two responsibilities in this method: manage the execution of snapshot tests (lines 97 to 116) and manage the execution of integration tests (119 to 148). I suggest you extract them to two different methods.